### PR TITLE
chore: add client lib PG runner to tpcc benchmark

### PR DESCRIPTION
**Pull request by hengfengli**
Add a runner that uses client library with PostgreSQL dialect.

### CODE REVIEW

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/BenchmarkApplication.java [hunk 1]
```diff
@@ -118,10 +118,7 @@ public void run(String... args) throws Exception {
       }
 
       if (tpccConfiguration.isRunBenchmark()
-          && (tpccConfiguration.getBenchmarkRunner().equals(TpccConfiguration.PGADAPTER_JDBC_RUNNER)
-              || tpccConfiguration
-                  .getBenchmarkRunner()
-                  .equals(TpccConfiguration.SPANNER_JDBC_RUNNER))) {
+          && (TpccConfiguration.RUNNERS.contains(tpccConfiguration.getBenchmarkRunner()))) {
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/BenchmarkApplication.java]: nit: should we show a clear error message if an invalid benchmark runner name has been given? Now we just skip running any benchmarks.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class JavaClientBenchmarkRunner extends AbstractBenchmarkRunner {
+  private static final Logger LOG = LoggerFactory.getLogger(JavaClientBenchmarkRunner.class);
+  private final ThreadLocal<Spanner> spannerThreadLocal = new ThreadLocal<>();
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: `Spanner` instances are intended a long-lived shared objects. So we should not create a separate `Spanner` instance for each thread.
The same also applies to `DatabaseClient`, but that happens automatically if you use a single `Spanner` instance. Multiple calls to `Spanner.getDatabaseClient(DatabaseId)` will return the same `DatabaseClient` instance.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+      Statistics statistics,
+      TpccConfiguration tpccConfiguration,
+      PGAdapterConfiguration pgAdapterConfiguration,
+      SpannerConfiguration spannerConfiguration,
+      Metrics metrics) {
+    super(statistics, tpccConfiguration, pgAdapterConfiguration, spannerConfiguration, metrics);
+  }
+
+  void setup() throws SQLException, IOException {
+    Spanner spanner = createSpanner();
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: We should not do this. Instead, the entire benchmark should use a single, shared `Spanner` instance.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    spannerThreadLocal.set(spanner);
+    databaseClientThreadLocal.set(
+        spanner.getDatabaseClient(
+            DatabaseId.of(
+                spannerConfiguration.getProject(),
+                spannerConfiguration.getInstance(),
+                spannerConfiguration.getDatabase())));
+    transactionManagerThreadLocal.set(databaseClientThreadLocal.get().transactionManager());
+    Uninterruptibles.sleepUninterruptibly(Duration.ofMillis(2000L));
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: Why is this here?

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  void teardown() throws SQLException {
+    transactionManagerThreadLocal.get().close();
+    spannerThreadLocal.get().close();
+  }
+
+  void executeStatement(String sql) throws SQLException {
+    if (sql.equals("begin transaction")) {
+      transactionThreadLocal.set(transactionManagerThreadLocal.get().begin());
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:76]: Does this really work? As in: `TransactionManager` instances are not intended to execute multiple transactions. Executing `TransactionManager#begin()` multiple times using the same `TransactionManager` should normally not work.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+  void setup() throws SQLException, IOException {
+    Spanner spanner = createSpanner();
+    spannerThreadLocal.set(spanner);
+    databaseClientThreadLocal.set(
+        spanner.getDatabaseClient(
+            DatabaseId.of(
+                spannerConfiguration.getProject(),
+                spannerConfiguration.getInstance(),
+                spannerConfiguration.getDatabase())));
+    transactionManagerThreadLocal.set(databaseClientThreadLocal.get().transactionManager());
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: See also below: `TransactionManager` is intended to execute one transaction and then be disposed, not to execute multiple, consecutive transactions.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  Object[] paramQueryRow(String sql, Object[] params) throws SQLException {
+    ParametersInfo parametersInfo = convertPositionalParametersToNamedParameters(sql, '?', "$");
+    Statement.Builder builder = Statement.newBuilder(parametersInfo.sqlWithNamedParameters);
+    bindParams(builder, params);
+    Statement stmt = builder.build();
+
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    ResultSet resultSet = transactionThreadLocal.get().executeQuery(stmt);
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:121]: This does not actually execute the query. You will see a query execution time in the range of a few microseconds, or maybe even less, but that is because `executeQuery` does not actually execute the query. Instead, the query will only be executed when you call `ResultSet#next()`

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    Duration executionDuration = stopwatch.elapsed();
+    metrics.recordLatency(executionDuration.toMillis());
+
+    if (!resultSet.next()) {
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: This is where the query is actually executed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+  public ParametersInfo convertPositionalParametersToNamedParameters(
+      String sql, char paramChar, String queryParameterPrefix) {
+    final String namedParamPrefix = queryParameterPrefix;
+    StringBuilder named = new StringBuilder(sql.length() + countOccurrencesOf(paramChar, sql));
+    int index = 0;
+    int paramIndex = 1;
+    while (index < sql.length()) {
+      char c = sql.charAt(index);
+      if (c == paramChar) {
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:173]: This does not work if the SQL string contains for example comments or a string literal with a positional parameter.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+      if (string.charAt(i) == c) {
+        res++;
+      }
+    }
+    return res;
+  }
+
+  // If getAsObject() in Struct can be a public method, then we should use it
+  // instead of using the following approach.
+  Object getObject(Value value) {
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:198]: Instead of relying on exception handling, can't you just call `Value#getType()` and use that in a switch statement to get the correct type?

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+  List<Object[]> executeParamQuery(String sql, Object[] params) throws SQLException {
+    ParametersInfo parametersInfo = convertPositionalParametersToNamedParameters(sql, '?', "$");
+    List<Object[]> results = new ArrayList<>();
+
+    Statement.Builder builder = Statement.newBuilder(parametersInfo.sqlWithNamedParameters);
+    bindParams(builder, params);
+    Statement stmt = builder.build();
+
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    ResultSet resultSet = transactionThreadLocal.get().executeQuery(stmt);
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:224]: Here also: This does not actually execute the query, so the value that you are measuring is not relevant and will be in the 'few microseconds' range.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+    Duration executionDuration = stopwatch.elapsed();
+    metrics.recordLatency(executionDuration.toMillis());
+
+    while (resultSet.next()) {
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: The first call to `resultSet.next()` executes the query

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/BenchmarkApplication.java (hunk 1 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/BenchmarkApplication.java]: If you look at `else`, we actually have the following: 

```
LOG.error("Unknown benchmark runner option: " + tpccConfiguration.getBenchmarkRunner());
```

Updated to throw an exception.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 4 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: I looked at your old code: https://github.com/GoogleCloudPlatform/pgadapter/pull/1267/files#diff-0fdd0faac71de04c6b133638083332b7e9210d05921f1acb089a13e311f87647R61

Should I delete it? I thought you must have a reason to add this.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 5 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:76]: Actually, it works. I thought we can call `TransactionManager` to begin a new transaction and then call `transactionManagerThreadLocal.get().commit();` or `transactionManagerThreadLocal.get().rollback();` to complete the transaction. After that, we can call `begin` again to get a new transaction. 

From what you say, do we always need to create a new `TransactionManager` to begin a new transaction?

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 2 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: Updated `spanner` to be a shared object and removed `DatabaseClient`.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 3 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: Updated.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 4 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: Removed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 7 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:121]: Yeah, I forgot this. Updated this.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 9 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:173]: I looked at the original parser but that is way complicated for this benchmark. I think this simplified version works okay as all queries are hard-coded and we know the format of generated strings.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 8 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 11 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:224]: Updated.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 12 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 10 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:198]: Good point. I didn't know `Value#getType()`. Cannot use switch as it requires `constant expression required`. Updated.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 5 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:76]: Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 6 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java]: Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 9 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:173]: I tried to set `lock-scanned-ranges` to true which introduces some comments. This works fine. For `a string literal with a positional parameter`, I don't see why we would have such a case in the benchmark. This code just simply replace `?` to positional parameters with `$<num>` (later, also `@p<num>`).

Comment (olavloite) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:173]: If you are certain that the benchmark does not have any SQL strings (and will not in the future) that might contain:
1. Comments that contain question marks.
2. String literals that contain question marks.

Then this should be fine.

But we should at least add a comment here to call that out clearly and explain that it is a deliberate choice, as a new developer who investigates this code in 2 years from now without any further context, might wonder how this ever worked.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 10 above)
Comment (olavloite) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:198]: (Note for the future: You can call `value.getType().getTypeCode()` and use that in a switch statement)

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+  }
+
+  Object[] paramQueryRow(String sql, Object[] params) throws SQLException {
+    ParametersInfo parametersInfo = convertPositionalParametersToNamedParameters(sql, '?', "$");
+    Statement.Builder builder = Statement.newBuilder(parametersInfo.sqlWithNamedParameters);
+    bindParams(builder, params);
+    Statement stmt = builder.build();
+
+    ResultSet resultSet = transactionThreadLocal.get().executeQuery(stmt);
+    Stopwatch stopwatch = Stopwatch.createStarted();
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:120]: nit: If you want to have a fully apples-to-apples comparison, then this should be moved to before the `executeQuery(..)` call above. `executeQuery(..)` does not actually execute the query, but it can include checking out the session from the pool. This is also always included in the `executeQuery(..)` methods in the JDBC driver and PGAdapter.

**Review by olavloite — APPROVED** [review]

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 9 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:173]: Good point. Added a comment.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java (hunk 13 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java:120]: Good catch. Fixed.
